Add advanced CORS headers

32bitkid opened the pull request:
This PR introduces an Advanced object property on Handlers, to define CORS related headers. These headers are added to _all_ requests, but are namely valuable when dealing with browser AJAX requests that send a preflight `OPTIONS` request before the _actual_ request.

Commit:
add cors headers to OPTIONS request.

Commit:
add cors headers to all requests

mattdodge commented on web_handler_block.py:
```diff
@@ -26,6 +46,11 @@ class WebHandler(GeneratorBlock):
 
+    cors = ObjectProperty(CORS,
+                          title="Access-Control Headers",
+                          default=dict(),
```
I think this needs to be `default=CORS()` for object property defaults.

32bitkid replied:
i was going to ask about this, wasn't sure what it was supposed to be.

mattdodge reviewed (changes requested):
I'm just now realizing that when a PR comes in from a fork our nio-blocks organization webhook doesn't trigger a test run. Dumb.

@32bitkid can you add an assertion that block properties make their way onto the response headers and also that the current tests still pass?

I think this PR is fine for a quick fix but ideally/eventually I'd like to see stronger CORS support from the web module interface. Allowing custom headers/origins/etc from the block config makes sense, but I think it should connect better to the core/module's CORS implementation. As it stands with this PR we are basically clobbering and ignoring the core CORS settings by default. This makes it not possible to do instance-wide CORS configuration and also adds to the already numerous places where we have CORS functionality defined.

mattdodge commented:
Also, for future reference, these values can be set at the instance level in your nio.conf like so:
```ini
[web]
cors_allow_origin = *
cors_allow_methods = GET,POST,PUT,DELETE,OPTIONS
cors_allow_headers = Accept, Origin, Content-Type, Authorization
cors_allow_credentials = true
```

Commit:
correct default value

Commit:
rearrange arguments to be less intrustive

move property evaulation into the block where it probably belongs.

Commit:
rethinking default parameters

Commit:
add tests

Commit:
guard on any falsy string value

32bitkid commented:
@mattdodge 
I implemented those tests along with some other changes/refactorings:

- re-ordered some the parameters to make it less intrusive to the existing tests/implementations. 
- refactored where the `cors` property gets evaluated, it happens in the block now, rather than in the handler, which feels like a better separation of concerns.
- changed some of the defaults to `None` so they will _not_ get overridden/added if they are empty, by default. The down-ish side of this is I don't think the system designer really support un-setting/`None`-ifying a property once it's been set, so I am guarding over a "falsey" string eval. Though that means you cant set a header to an empty value. But I think that might be okay.
- because these WebHandlers "inherit" from the global settings, it means it's not really feasible to "unset" a header. For example the `Access-Control-Allow-Credentials` header should only ever have a value of `true`. If you want to set it to "false", you should _remove_ the header from the response. 

    It might be good if these did _not_ inherit headers from the global web context. but that feels like a pretty big change that we'd need to talk through. I don't think its a big deal for now, but something to consider in the future, if decoupling these "application-layer" handlers from the "core-layer" handlers. Or at least, talk through a better model of the relationship between the different types of Handlers, between those that expose instance/service/block APIs and those that expose application-level concerns.

mattdodge commented on handler.py:
```diff
@@ -96,6 +100,10 @@ def validate_method(self, method, rsp):
 
+    def __add_headers(self, rsp):
+        if self._headers is not None:
+            for header in self._headers:
```
For future reference, no change needed;
```python
for header_name, header_value in self._headers.items():
    rsp.set_header(header_name, header_value)
```
